financial-templates-lib/Logger.js

```diff
@@ -28,50 +28,87 @@ const alignedWithColorsAndTime = winston.format.combine(
   })
 );
 
+const slackFormatter = info => {
+  // Each part of the slack response is a separate block with markdown text within it.
+  // All slack responses start with the heading level and where the message came from.
+  let formattedResponse = {
+    blocks: [
+      {
+        type: "section",
+        text: {
+          type: "mrkdwn",
+          text: `*${info.level}: ${info.at}* --> ${info.message}`
+        }
+      }
+    ]
+  };
+
+  // All messages from winston come in as a Json object. The loop below expands this object
+  //and adds mrkdwn sections for each key value pair with a bullet point. If the section is
+  // an object then it was passed containing multiple sub points. This is also expanded as a
+  // sub indented section. If the key is `tx` then it is encoded as a etherscan URL.
+  for (const key in info) {
+    // these keys have been printed in the previous block.
+    if (key == "at" || key == "level" || key == "message") {
+      continue;
+    }
+    if (typeof info[key] === "object" && info[key] !== null) {
+      formattedResponse.blocks.push({
+        type: "section",
+        text: {
+          type: "mrkdwn",
+          text: ` • _${key}_:\n`
+        }
+      });
+      for (const subKey in info[key]) {
+        if (subKey == "tx") {
+          formattedResponse.blocks[
+            formattedResponse.blocks.length - 1
+          ].text.text += `    - _tx_: <https://etherscan.io/tx/${info[key][subKey]}|${info[key][subKey]}> \n`;
+        } else {
+          formattedResponse.blocks[
+            formattedResponse.blocks.length - 1
+          ].text.text += `    - _${subKey}_: ${info[key][subKey]}\n`;
+        }
+      }
+    } else {
```

Comment:
What goes through this else block? E.g., if info[key] == null, what would we print in that case?

Reply:
The `if (typeof info[key] === "object" && info[key] !== null) {` is used to check if the prop in the key value paring is an object. if it is, it is spread out and delimited with the `-` punctuation mark. If it's not an object, then it is a string or int and is delimited with `•`.

Therefore this `else` block is entered for all non-objects that are in the winston message.